languages/tree-sitter-stack-graphs-typescript/build.rs

```diff
@@ -0,0 +1,73 @@
+// -*- coding: utf-8 -*-
+// ------------------------------------------------------------------------------------------------
+// Copyright © 2024, stack-graphs authors.
+// Licensed under either of Apache License, Version 2.0, or MIT license, at your option.
+// Please see the LICENSE-APACHE or LICENSE-MIT files in this distribution for license details.
+// ------------------------------------------------------------------------------------------------
+
+use std::path::Path;
+
+use anyhow::{Result, bail};
+use regex::Regex;
+
+const TSG_SOURCE: &str = "src/stack-graphs.tsg";
+const DIALECTS: [&str; 2] = ["typescript", "tsx"];
+
+/// preprocess the input file, removing lines that are not for the selected dialect
+fn preprocess(input: impl std::io::Read, mut output: impl std::io::Write, dialect: &str) -> Result<()> {
+    // Matches:   ; #dialect typescript
+    let directive_start = Regex::new(r";\s*#dialect\s+(\w+)").unwrap();
+
+    // Matches:   ; #end
+    let directirve_end = Regex::new(r";\s*#end").unwrap();
+
+    let input = std::io::read_to_string(input)?;
+
+    // If the filter is None or Some(true), the lines are written to the output
+    let mut filter: Option<bool> = None;
+
+    for (line_no, line) in input.lines().enumerate() {
+        if let Some(captures) = directive_start.captures(line) {
+            let directive = captures.get(1).unwrap().as_str();
+            if !DIALECTS.contains(&directive) {
+                bail!("Line {line_no}: unknown dialect: {directive}");
+            }
+
+            filter = Some(dialect == directive);
+            output.write_all(line.as_bytes())?;
```

Comment:
I would keep the directives out of the generated file where they have already been applied.

```suggestion
```

Reply:
I did this for two reasons:
1. Since I process a full line at a time there's a chance that a comment / directive follows valid TSG code:
```
(function) { ...} ;  #dialect tsx
```
always writing out the line ensures that the valid code is not lost.

2. When looking at the generated code, I found it convenient to still see the directives as they highlight dialect specific sections.

Reply:
My thinking was that in the generated code the the directives are already processed, and them not being there makes that obvious.

I see your point about the comment coming after some other code. (And this is always the trouble with text-based preprocessing instead of something properly supported in the language---but that's the cost of going for an expedient solution here ;).) My expectation with markers in pairs like this is that they operate on blocks of lines. A tricky example with keeping the lines would be:

   ; #dialect tsx
   (function) {
     ...
   } ; #end

If this is processed with the current algorithm, it would leave an umatched `}` I think. I think treating them as operating on lines simplifies things. If you're worried that people may mistakes, you can add an extra check that the line before the matched directive matches `[\s;]+`, or throw an error that a directive cannot be on the same line as code.

Reply:
you're correct that I accounted for one edge case but not the mirror image of it. I'll add a check to ensure that directives are not mixed with code. And then I won't output the directives.